File: SINGLYLL.py
##############################################################################
# Application Name: SINGLYLL
# Description: It is a Singly linear linked list (Data structure) application.
# Date: 05/10/2021.
# Author Name: Harish Vijay Bavne
##############################################################################

# Node Declaration
class node:
    def __init__(self,no):
        self.Data = no
        self.next = None
        
# Class for Singly linear Linked list
class SinglyLL:
    # class variable private for node count
    __Size = 0                      
    
    def __init__(self):
        # Head is for first node reference
        self.Head = None            
        
        
    # Function for Insert elements in linked list at first position
    def InsertFirst(self,No):       
        newn = node(No)
        if (self.Head == None):
            self.Head = newn
            SinglyLL.__Size = SinglyLL.__Size + 1
        else:
            newn.next = self.Head
            self.Head = newn
            SinglyLL.__Size = SinglyLL.__Size + 1
    
    # Function for Insert element in linked list at last position
    def InsertLast(self,No):
        newn = node(No)
        if (self.Head == None):
            self.Head = newn
            SinglyLL.__Size = SinglyLL.__Size + 1
        else:
            temp = self.Head
            while(temp.next != None):
                temp = temp.next
            temp.next = newn    
            SinglyLL.__Size = SinglyLL.__Size + 1
    
    # Function for Insert element in linked list at said position
    def InsertAtPosition(self,No,Pos):
        newn = node(No)
        temp = self.Head
        iSize = SinglyLL.__Size
        if ((Pos < 1) or (Pos > iSize+1)):
            return
        if (Pos == 1):
            self.InsertFirst(No)
        elif (Pos == iSize+1):
            self.InsertLast(No)
        else:    
            print("P")
            for i in range(1,Pos-1):
                temp = temp.next
            newn.next = temp.next
            temp.next = newn
            SinglyLL.__Size = SinglyLL.__Size + 1
    
    # Function for Delete element in Linked list at first position
    def DeleteFirst(self):
        temp = self.Head
        if (self.Head == None):
            return
        elif (self.Head.next == None):
            del(temp)
            self.Head = None
            SinglyLL.__Size = SinglyLL.__Size - 1
        else:
            self.Head = temp.next
            del(temp)
            SinglyLL.__Size = SinglyLL.__Size - 1
    
    # Function for Delete element in Linked list at Last position
    def DeleteLast(self):
        temp = self.Head
        if (self.Head == None):
            return
        elif (self.Head.next == None):
            del(temp)
            self.Head = None
            SinglyLL.__Size = SinglyLL.__Size - 1
        else:
            for i in range(1,SinglyLL.__Size-1):
                temp = temp.next
            del(temp.next)
            temp.next = None
            SinglyLL.__Size = SinglyLL.__Size - 1
    
    # Function for Delete element at said position in Linked list
    def DeleteAtPosition(self,Pos):
        temp =self.Head
        iSize = SinglyLL.__Size
        target = None
        if ((Pos < 1)or(Pos > iSize)):
            return
        elif (Pos == 1):
            self.DeleteFirst()
        elif (Pos == iSize):
            self.DeleteLast()
        else:
            for i in range(1,Pos-1):
                temp = temp.next
            target = temp.next
            temp.next = target.next
            del(target)
            SinglyLL.__Size = SinglyLL.__Size - 1
    
    # Function for Display elements in linked list
    def Display(self):
        temp = self.Head
        while(temp != None):
            print(f"|{temp.Data}|=>",end="")
            temp = temp.next
        print("None")
        
    # Function for count number of Nodes in linked list
    def Count(self):
        return SinglyLL.__Size
    
# Entry point function main.
def main():
    obj = SinglyLL()
    def menu():
        print("*********************************")
        print("1:Insert First")
        print("2:Insert last")
        print("3:Insert at Position")
        print("4:DeleteFirst ")
        print("5:DeleteLast")
        print("6:DeleteAtPosition")
        print("7:Display element")
        print("8:Count Number of Nodes")
        print("0:Exit application")
        print("*********************************")
    menu()
    choice = int(input("Enter a your choice:"))
    while True:
        if (choice == 1):
            No = int(input("Enter Element:"))
            obj.InsertFirst(No)
        elif (choice == 2):
            No = int(input("Enter Element:"))
            obj.InsertLast(No)
        elif (choice == 3):
            No = int(input("Enter Element:"))
            Pos = int(input("Enter position:"))
            obj.InsertAtPosition(No,Pos)
        elif (choice == 4):
            obj.DeleteFirst()
        elif (choice == 5):
            obj.DeleteLast()
        elif (choice == 6):
            Pos = int(input("Enter a position:"))
            obj.DeleteAtPosition(Pos)
        elif (choice == 7):
            obj.Display()
        elif (choice == 8):
            Ret = obj.Count()
            print(f"Number of Nodes are:{Ret}")
        elif (choice == 0):
            print("Thank you for using application")
            break
        else:
            print("Enter a correct choice.")
        menu()
        choice = int(input("Enter a your choice:"))
# Starter
if __name__ == "__main__":
    main()

File: SINGLYLL.cpp

#include<iostream>
using namespace std;

struct node
{
	int Data;
	struct node *next;
};

typedef struct node NODE;
typedef struct node* PNODE;
class SinglyLL
{
	private:
		int Size;
		PNODE Head;
	public:
	SinglyLL()
	{
		Size = 0;
		Head = NULL;
	}
	void InsertFirst(int no)
	{
		PNODE newn = new NODE;
		newn->Data = no;
		newn->next = NULL;
		if (Head == NULL)
		{
			Head = newn;
			Size ++;
		}
		else
		{
			newn->next = Head;
			Head = newn;
			Size ++;
		}
	}
	void InsertLast(int no)
	{
		PNODE temp = Head;
		PNODE newn = new NODE;
		newn->Data = no;
		newn->next = NULL;
		if (Head == NULL)
		{
			Head = newn;
			Size++;
		}
		else
		{
			while(temp->next != NULL)
			{
				temp = temp->next;
			}
			temp->next = newn;
			Size++;
		}
	}
	
	void InsertAtPosition(int no,int pos)
	{
		PNODE temp = Head;
		PNODE newn = new NODE;
		newn->Data = no;
		newn->next = NULL;
		
		if ((pos <1)||(pos>Size+1))
		{
			return;
		}
		else if (pos == 1)
		{
			InsertFirst(no);
		}
		else if(pos == Size+1)
		{
			InsertLast(no);
		}
		else
		{
			for (int i=1;i<pos-1;i++)
			{
				temp = temp->next;
			}
			newn->next = temp->next;
			temp->next = newn;
			Size++;
		}
	}
	
	void DeleteFirst()
	{
		PNODE temp = Head;
		if(Head == NULL)
		{
			return;
		}
		if (Head->next == NULL)
		{
			delete(Head);
			Head = NULL;
			Size--;
		}
		else
		{
			Head = temp->next;
			delete(temp);
			Size--;
		}
	}
	
	void DeleteLast()
	{
		PNODE temp = Head;
		if(Head == NULL)
		{
			return;
		}
		if (Head->next == NULL)
		{
			delete(Head);
			Head = NULL;
			Size--;
		}
		else
		{
			while(temp->next->next != NULL)
			{
				temp = temp->next;
			}
			delete(temp->next);
			temp->next = NULL;
			Size--;
			
		}
	}
	
	void DeleteAtPosition(int Pos)
	{
		PNODE target=NULL;
		PNODE temp = Head;
		if ((Pos < 1)||(Pos > Size))
		{
			return;
		}
		else if(Pos == 1)
		{
			DeleteFirst();
		}
		else if(Pos == Size)
		{
			DeleteLast();
		}
		else
		{
			for (int i = 1;i<(Pos-1);i++)
			{
				temp = temp->next;
			}
			target = temp->next;
			temp->next = target->next;
			delete(target);
			Size--;
		}
	}
	void Display()
	{
		PNODE temp = Head;
		while(temp != NULL)
		{
			cout<<"|"<<temp->Data<<"|=>";
			temp = temp->next;
		}
		cout<<"\n";
	}
	int Count()
	{
		return Size;
	}
};

int main()
{
	SinglyLL obj;
	int Value = 0,Pos = 0,choice = 1,ret = 0;
	while(choice !=0)
	{
		cout<<"\n==============================\n";
		cout<<"SINGLY LINKED LIST\n";
		cout<<"1:InsertFirst \n";
		cout<<"2:InsertLast \n";
		cout<<"3:InsertAtPosition \n";
		cout<<"4:DeleteFirst \n";
		cout<<"5:DeleteLast \n";
		cout<<"6:DeleteAtPosition \n";
		cout<<"7:Display \n";
		cout<<"8:Count \n";
		cout<<"9:Exit \n";
		cout<<"\n==============================\n";
		cout<<"Please enter your choice:";
		cin>>choice;
		
		switch(choice)
		{
			case 1:
			{
				cout<<"Enter a element:";
				cin>>Value;
				obj.InsertFirst(Value);
				break;
			}
			case 2:
			{
				cout<<"Enter a element:";
				cin>>Value;
				obj.InsertLast(Value);
				break;
			}
			case 3:
			{
				cout<<"Enter a element:";
				cin>>Value;
				cout<<"Enter a position:";
				cin>>Pos;
				obj.InsertAtPosition(Value,Pos);
				break;
			}
			case 4:
			{
				obj.DeleteFirst();
				break;
			}
			case 5:
			{
				obj.DeleteLast();
				break;
			}
			case 6:
			{
				cout<<"Enter position:";
				cin>>Pos;
				obj.DeleteAtPosition(Pos);
				break;
			}
			case 7:
			{
				obj.Display();
				break;
			}
			case 8:
			{
				ret = obj.Count();
				cout<<"Number of nodes in linked list:"<<ret<<"\n";
				break;
			}
			case 9:
			{
				cout<<"Thank your for using application\n";
				return 0;
			}
			default:
			{
				cout<<"Please eneter valid choice";
				break;
			}
		}
	}
	
	return 0;
	
}